File: flutter/ai_app_bridge_flutter/android/src/main/kotlin/io/github/mobileaidev/aiappbridge/flutter/AiAppBridgeFlutterPlugin.kt
package io.github.mobileaidev.aiappbridge.flutter

import android.app.Activity
import android.content.Context
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class AiAppBridgeFlutterPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    private var channel: MethodChannel? = null
    private var activity: Activity? = null
    private var applicationContext: Context? = null
    private var bridgeActionCallback: Any? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, channelName).also {
            it.setMethodCallHandler(this)
        }
        startBridge(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        unregisterFlutterActionBridge()
        channel?.setMethodCallHandler(null)
        channel = null
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        startBridge(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        unregisterFlutterActionBridge()
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        unregisterFlutterActionBridge()
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (call.method == "checkAction") {
                val bridgeClass = Class.forName(androidBridgeClassName)
                result.success(bridgeClass.getMethod("checkFlutterAction", String::class.java)
                    .invoke(null, call.arguments as String))
                return
            }
            val handled = when (call.method) {
                "updateSnapshot" -> updateSnapshot(call.arguments?.toString() ?: "{}")
                "recordLog", "recordNetwork", "recordState", "recordEvent" ->
                    recordCapture(call.method, call.arguments as String)
                else -> {
                    result.notImplemented()
                    return
                }
            }
            result.success(
                if (handled) {
                    mapOf("ok" to true)
                } else {
                    mapOf("ok" to false, "reason" to "debug_bridge_absent")
                },
            )
        } catch (error: Throwable) {
            result.error("AI_APP_BRIDGE_CALL_FAILED", error.message, null)
        }
    }

    private fun startBridge(context: Context) {
        try {
            val bridgeClass = Class.forName(androidBridgeClassName)
            val startMethod = bridgeClass.getMethod("start", Context::class.java)
            startMethod.invoke(null, context)
            // A background service also creates an engine and registers plugins.
            // Only an engine attached to an Activity owns the UI action channel.
            if (context is Activity) registerFlutterActionBridge(bridgeClass)
        } catch (_: ClassNotFoundException) {
            // The Android SDK is optional so this Flutter plugin can stay reusable.
        } catch (error: Throwable) {
            Log.w(tag, "failed to start Android AI app bridge", error)
        }
    }

    private fun registerFlutterActionBridge(bridgeClass: Class<*>) {
        if (bridgeActionCallback != null) {
            return
        }
        val currentChannel = channel ?: return
        val callbackClass = Class.forName("$androidBridgeClassName\$FlutterActionHandler")
        val callback = java.lang.reflect.Proxy.newProxyInstance(
            callbackClass.classLoader,
            arrayOf(callbackClass),
        ) { _, method, args ->
            if (method.name == "handle") {
                val values = requireNotNull(args)
                val reply = values[2]
                val replyMethod = Class.forName("$androidBridgeClassName\$FlutterActionReply").getMethod("reply", String::class.java)
                runFlutterAction(currentChannel, values[0] as String, values[1] as String) { response -> replyMethod.invoke(reply, response) }
                null
            } else {
                null
            }
        }
        bridgeClass.getMethod("setFlutterActionHandler", callbackClass).invoke(null, callback)
        bridgeActionCallback = callback
    }

    private fun unregisterFlutterActionBridge() {
        val callback = bridgeActionCallback ?: return
        try {
            val bridgeClass = Class.forName(androidBridgeClassName)
            bridgeClass.getMethod("clearFlutterActionHandler", Class.forName("$androidBridgeClassName\$FlutterActionHandler"))
                .invoke(null, callback)
        } catch (_: Throwable) {
        } finally {
            bridgeActionCallback = null
        }
    }

    private fun runFlutterAction(currentChannel: MethodChannel, method: String, payloadJson: String, reply: (String) -> Unit) {
        val posted = android.os.Handler(android.os.Looper.getMainLooper()).post {
            try { currentChannel.invokeMethod(
                method,
                payloadJson,
                object : MethodChannel.Result {
                    override fun success(result: Any?) {
                        reply(toJsonString(result))
                    }

                    override fun error(errorCode: String, errorMessage: String?, errorDetails: Any?) {
                        reply(JSONObject().put("ok", false).put("error", errorCode).put("message", errorMessage).toString())
                    }

                    override fun notImplemented() {
                        reply("""{"ok":false,"error":"flutter_action_not_implemented"}""")
                    }
                },
            ) } catch (error: Throwable) {
                reply(JSONObject().put("ok", false).put("error", "flutter_channel_failed").put("message", error.message).toString())
            }
        }
        if (!posted) reply("""{"ok":false,"error":"flutter_main_thread_unavailable"}""")
    }

    private fun toJsonString(value: Any?): String {
        return when (value) {
            null -> """{"ok":false,"error":"empty_flutter_action_result"}"""
            is String -> value
            is Map<*, *> -> JSONObject(value).toString()
            else -> JSONObject()
                .put("ok", false)
                .put("error", "invalid_flutter_action_result")
                .toString()
        }
    }

    private fun updateSnapshot(snapshotJson: String): Boolean {
        val bridgeClass = try {
            Class.forName(androidBridgeClassName)
        } catch (_: ClassNotFoundException) {
            return false
        }
        val updateMethod = bridgeClass.getMethod("updateFlutterSnapshot", String::class.java)
        updateMethod.invoke(null, snapshotJson)
        return true
    }

    private fun recordCapture(method: String, payloadJson: String): Boolean {
        return invokeAndroidBridge("recordFlutterCapture") { bridgeClass ->
            bridgeClass.getMethod("recordFlutterCapture", String::class.java, String::class.java)
                .invoke(null, method, payloadJson)
        }
    }

    private fun invokeAndroidBridge(methodName: String, call: (Class<*>) -> Unit): Boolean {
        val bridgeClass = try {
            Class.forName(androidBridgeClassName)
        } catch (_: ClassNotFoundException) {
            return false
        }
        try {
            call(bridgeClass)
            return true
        } catch (error: Throwable) {
            Log.w(tag, "failed to call Android AI app bridge method $methodName", error)
            throw error
        }
    }

    companion object {
        private const val tag = "AiAppBridge"
        private const val channelName = "ai_app_bridge"
        private const val androidBridgeClassName = "io.github.mobileaidev.aiappbridge.android.AiAppBridge"
    }
}

File: flutter/ai_app_bridge_flutter/ios/ai_app_bridge_flutter/Sources/ai_app_bridge_flutter/AiAppBridgeFlutterPlugin.swift
import Flutter
import Foundation
import UIKit

#if canImport(AiAppBridgeIOS)
import AiAppBridgeIOS
#endif

public final class AiAppBridgeFlutterPlugin: NSObject, FlutterPlugin {
    private static let channelName = "ai_app_bridge"
    private var channel: FlutterMethodChannel?
    private var handlerToken: String?

    private init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = AiAppBridgeFlutterPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)

        AiAppBridge.shared.start()
        instance.handlerToken = AiAppBridge.shared.setFlutterActionHandler { [weak instance] method, payload, reply in
            guard let instance else {
                reply(#"{"ok":false,"error":"flutter_plugin_detached"}"#); return
            }
            instance.runFlutterAction(method: method, payloadJson: payload, reply: reply)
        }
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        if let token = handlerToken { AiAppBridge.shared.clearFlutterActionHandler(token: token) }
        handlerToken = nil
        channel = nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkAction":
            guard let payload = call.arguments as? String else {
                result(#"{"ok":false,"error":"invalid_flutter_execution_identity"}"#); return
            }
            result(AiAppBridge.shared.checkFlutterAction(payload))
        case "updateSnapshot":
            guard let payload = call.arguments as? String else {
                result(["ok": false, "error": "invalid_snapshot_argument"])
                return
            }
            AiAppBridge.shared.updateFlutterSnapshot(payload)
            result(["ok": true])
        case "recordLog", "recordNetwork", "recordState", "recordEvent":
            guard let payload = call.arguments as? String else {
                result(["ok": false, "error": "invalid_capture_argument"])
                return
            }
            AiAppBridge.shared.recordFlutterCapture(method: call.method, payloadJson: payload) { response in
                DispatchQueue.main.async { result(response) }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func runFlutterAction(method: String, payloadJson: String, reply: @escaping (String) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let channel = self?.channel else {
                reply(#"{"ok":false,"error":"flutter_channel_absent"}"#); return
            }
            channel.invokeMethod(method, arguments: payloadJson) { value in
                if let error = value as? FlutterError {
                    reply(Self.jsonString(["ok": false, "error": error.code,
                                           "message": error.message as Any? ?? NSNull()])); return
                }
                guard let result = value as? [String: Any] else {
                    reply(#"{"ok":false,"error":"invalid_flutter_action_result"}"#); return
                }
                reply(Self.jsonString(result))
            }
        }
    }

    private static func jsonString(_ value: [String: Any]) -> String {
        guard JSONSerialization.isValidJSONObject(value),
              let data = try? JSONSerialization.data(withJSONObject: value),
              let text = String(data: data, encoding: .utf8) else {
            return #"{"ok":false,"error":"invalid_flutter_action_result"}"#
        }
        return text
    }
}
